pkg/sinks/kafka/kafka.go

```diff
@@ -102,37 +96,50 @@ func (tk *ToKafka) GetName() string {
 // Write writes to the kafka topic.
 func (tk *ToKafka) Write(_ context.Context, messages []isb.Message) ([]isb.Offset, []error) {
 	errs := make([]error, len(messages))
-	wg := new(sync.WaitGroup)
-
-	sinkCh := make(chan *sinkMessage)
-
-	for i := uint32(0); i < tk.concurrency; i++ {
-		wg.Add(1)
-		go func(msgCh chan *sinkMessage) {
-			defer wg.Done()
-			for message := range msgCh {
-				_, _, err := tk.producer.SendMessage(message.message)
-				if err != nil {
-					kafkaSinkWriteErrors.With(map[string]string{"vertex": tk.name, "pipeline": tk.pipelineName}).Inc()
-					tk.log.Errorw("SendMessage failed", zap.Error(err), zap.Any("message", message))
-				} else {
-					kafkaSinkWriteCount.With(map[string]string{"vertex": tk.name, "pipeline": tk.pipelineName}).Inc()
-				}
-				//keep error in message index
-				errs[message.index] = err
-			}
-		}(sinkCh)
+	for i := 0; i < len(errs); i++ {
+		errs[i] = fmt.Errorf("unknown error")
 	}
-	for idx, message := range messages {
-		// producer.Produce can be changed in to producer.produceBatch later once API is ready for production scale
+	done := make(chan struct{})
+	timeout := time.After(3 * time.Second)
+	go func() {
+		sent := 0
+		for {
+			if sent == len(messages) {
+				close(done)
+				return
+			}
+			select {
+			case err := <-tk.producer.Errors():
+				idx := err.Msg.Metadata.(int)
+				errs[idx] = err.Err
+				sent++
+			case m := <-tk.producer.Successes():
+				idx := m.Metadata.(int)
+				errs[idx] = nil
+				sent++
+			case <-timeout:
```

Comment:
There will be problems here, need more stuff...

Reply:
can you send a counter metric and also a log line so we can track these timeouts?

Reply:
done.